lib/src/drawing_and_graphics/shadow.dart

```diff
@@ -0,0 +1,129 @@
+import 'dart:ui';
+
+import 'package:flutter/widgets.dart';
+
+/// A widget that adds a shadow effect to its child widget.
+class Shadow extends StatelessWidget {
+  /// Creates a Shadow widget with the given properties.
+  ///
+  /// Use the [radius] to set the amount of blur, where `0.0` is sharp and a higher
+  /// value is more fuzzy. You can also set the shadow [color] and ([x], [y]) offset.
+  const Shadow({
+    super.key,
+    // In SwiftUI the default is `Color(.sRGBLinear, white: 0, opacity: 0.33)`
+    this.color = const Color(0x54000000),
+    required this.radius,
+    this.x = 0,
+    this.y = 0,
+    required this.child,
+  });
+
+  /// The color of the shadow.
+  final Color color;
+
+  /// The blur radius of the shadow.
+  ///
+  /// `0.0` is sharp while higher values are more fuzzy.
+  final double radius;
+
+  /// The horizontal offset of the shadow.
+  final double x;
+
+  /// The vertical offset of the shadow.
+  final double y;
+
+  /// The widget to which the shadow effect will be applied.
+  final Widget child;
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      children: [
+        Transform.translate(
+          offset: Offset(x, y),
+          child: Blur(
+            sigmaX: radius,
+            sigmaY: radius,
+            child: Tint(
+              color: color,
+              child: Desaturate(
```

Comment:
I wonder if this chain of adjustments would benefit from a custom shader. We're removing color from every pixel, then applying color to every pixel, then blurring the overall shape. Maybe these widgets' shaders are already optimized in the engine. But also perhaps a single shader that does this in one step would be more performant...

Reply:
Perhaps. If we were only thinking about this package, I'd say let's wait to see until we actually observe a performance problem. However, it might be interesting to implement a custom shader. I'm willing to give it a try.

Reply:
It's not a big deal for now, especially if we're gonna make those widgets private. But it could be a neat experiment if you wanna play with it.

Reply:
I've been traveling these days and haven't had much chance to work on this, but when I do I'll take a look at shaders.

Reply:
I've been playing around with shaders for a few days. It's been interesting learning about them, but I haven't gotten a version that was better than the pure Flutter version that I had before. Desaturation and blur is partially working but the algorithm is really slow for high blur values. I'm going to stop trying for now but post the code here in case I decide to try again in the future.

shadow.frag

```
#include <flutter/runtime_effect.glsl>

uniform float desaturationFactor;
uniform float width;
uniform float height;
uniform float blurStrength;
uniform sampler2D image;

out vec4 fragColor;

vec4 sampleOffset(vec2 uv, vec2 offset) {
  return texture(image, uv + offset / vec2(width, height));
}

void main() {
  vec2 uv = FlutterFragCoord().xy / vec2(width, height);
  
  vec4 blurred = vec4(0.0);
  float total = 0.0;
  for (float x = -20.0; x <= 20.0; x++) {
    for (float y = -20.0; y <= 20.0; y++) {
      float weight = exp(-(x*x + y*y) / (2.0 * blurStrength * blurStrength));
      blurred += sampleOffset(uv, vec2(x, y)) * weight;
      total += weight;
    }
  }
  blurred /= total;
  
  float gray = dot(blurred.rgb, vec3(0.299, 0.587, 0.114));
  vec3 desaturated = mix(blurred.rgb, vec3(gray), desaturationFactor);
  
  fragColor = vec4(desaturated, blurred.a);
}
```

shadow.dart

```
import 'package:flutter/material.dart';
import 'package:flutter_shaders/flutter_shaders.dart';

class Shadow extends StatelessWidget {
  final Widget child;
  final double desaturationFactor;

  const Shadow({
    super.key,
    required this.child,
    this.desaturationFactor = 1.0,
  });

  @override
  Widget build(BuildContext context) {
    return ShaderBuilder(
      assetKey: 'assets/shaders/shadow.frag',
      (context, shader, child) => AnimatedSampler(
        (image, size, canvas) {
          shader
            ..setFloat(0, desaturationFactor)
            ..setFloat(1, size.width)
            ..setFloat(2, size.height)
            ..setFloat(3, 10.0)
            ..setImageSampler(0, image);

          canvas.drawRect(
            Offset.zero & size,
            Paint()..shader = shader,
          );
        },
        child: Padding(
          padding: const EdgeInsets.all(50),
          child: child!,
        ),
      ),
      child: child,
    );
  }
}
```

main.dart

```
import 'package:flutter/material.dart';
import 'package:shader_test/shadow.dart';

Future<void> main() async {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Shadow(
            child: Container(
              width: 200,
              height: 200,
              color: Colors.amber,
              child: const Text(
                'Hello World',
                style: TextStyle(color: Colors.red),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

<img width="272" alt="Screenshot 2024-07-25 at 09 58 20" src="https://github.com/user-attachments/assets/887ee642-1c3d-435e-a2eb-e717e0be2160">